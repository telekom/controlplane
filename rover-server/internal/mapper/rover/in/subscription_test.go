// Copyright 2025 Deutsche Telekom IT GmbH
//
// SPDX-License-Identifier: Apache-2.0

package in

import (
	"github.com/gkampitakis/go-snaps/snaps"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	roverv1 "github.com/telekom/controlplane/rover/api/v1"

	"github.com/telekom/controlplane/rover-server/internal/api"
)

var _ = Describe("Subscription Mapper", func() {
	Context("MapApiSubscription", func() {
		It("must map ApiSubscription correctly", func() {
			input := apiSubscription

			output := mapApiSubscription(input)

			Expect(output).ToNot(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must handle empty ApiSubscription input", func() {
			input := api.ApiSubscription{}

			output := mapApiSubscription(input)

			Expect(output).ToNot(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})
	})

	Context("mapSubscription", func() {
		It("must map an ApiSubscription correctly", func() {
			input := GetApiSubscription(apiSubscription)
			output := &roverv1.Subscription{}

			err := mapSubscription(&input, output)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must map an EventSubscription correctly", func() {
			input := GetEventSubscription(eventSubscription)
			output := &roverv1.Subscription{}

			err := mapSubscription(&input, output)

			Expect(err).To(BeNil())
			snaps.MatchSnapshot(GinkgoT(), output)
		})

		It("must return an error if Discriminator fails", func() {
			input := &api.Subscription{}
			output := &roverv1.Subscription{}

			err := mapSubscription(input, output)

			Expect(err).ToNot(BeNil())
			Expect(err.Error()).To(ContainSubstring("failed to get subscription type"))
			snaps.MatchSnapshot(GinkgoT(), output)
		})
	})

})
