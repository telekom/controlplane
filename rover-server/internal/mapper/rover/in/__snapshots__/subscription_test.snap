
[Subscription Mapper MapApiSubscription must map ApiSubscription correctly - 1]
&v1.ApiSubscription{
    BasePath:       "/eni/distr/v1",
    Organization:   "",
    Transformation: (*v1.Transformation)(nil),
    Traffic:        v1.SubscriberTraffic{},
    Security:       (*v1.SubscriberSecurity)(nil),
}
---

[Subscription Mapper MapApiSubscription must handle empty ApiSubscription input - 1]
&v1.ApiSubscription{}
---

[Subscription Mapper mapSubscription must map an ApiSubscription correctly - 1]
&v1.Subscription{
    Api: &v1.ApiSubscription{
        BasePath:       "/eni/distr/v1",
        Organization:   "",
        Transformation: (*v1.Transformation)(nil),
        Traffic:        v1.SubscriberTraffic{},
        Security:       (*v1.SubscriberSecurity)(nil),
    },
    Event: (*v1.EventSubscription)(nil),
}
---

[Subscription Mapper mapSubscription must map an EventSubscription correctly - 1]
&v1.Subscription{
    Api:   (*v1.ApiSubscription)(nil),
    Event: &v1.EventSubscription{EventType:"test-event"},
}
---

[Subscription Mapper mapSubscription must return an error if Discriminator fails - 1]
&v1.Subscription{}
---
